ci/closures.sh

```diff
@@ -1,12 +1,22 @@
 #!/bin/bash
-set -xe
+set -x
 set -o pipefail
 
 # Enables printing of all potential GC allocation sources to stdout
 export DFLAGS=-vgc
 
 # Prepare sources
-beaver dlang make d2conv
+beaver dlang make d2conv; test $? -eq 0 || exit 2
 
-# Ensure that there are no lines about closure allocations in the output
-! beaver dlang make fasttest 2>&1 | grep "closure"
+# Ensure there are no lines about closure allocations in the output.
+# Note explicit check for `grep` exit status 1, i.e. no lines found.
+# This means tests will fail if any of these conditions are met:
+#
+#   * `make` fails (exit status 2, propagated by `set -o pipefail`)
+#   * `grep` fails (exit status 2)
+#   * `grep` finds `closure` (exit status 0)
+#
+# Note that `make` can in principle exit with status 1 too, but only
+# if the `-q` option is used, which should never be the case for our
+# builds, so we can assume exit status 1 only comes from `grep`.
+beaver dlang make fasttest 2>&1 | (grep -e "closure"; test $? -eq 1)
```

Comment:
I'd rather use something like named pipe and multi-line scirpt. For example (untested):

```suggestion
mkfifo output
beaver dlang make fasttest 2>&1 > output || exit 1
grep "closure" output
test $? -eq 1 || exit 1
```

Reply:
Out of curiosity, since you wrote the original script: why is the `2>&1` necessary/wanted?  I ask because the question came up with a team member in another context.  I assume it's in order to ensure that the test works regardless of whether the compiler outputs to `stdout` or `stderr`, but thought I'd ask.